### ISSUE TITLE
Shared package tech debts

### ISSUE THREAD

**Issue by danwt** [dymensionxyz/dymension#931]
The genesis bridge, when it goes in, will create a couple of tech debts:

- [ ] messy gerrs package hub used also in dymint
- [ ] duplicated json definitions in rdk and hub

Probably the RDK or the Hub would be a good place to move these definitions, which would introduce a dependency from repo to the other, but this is favoured by some on the team (as opposed to an entirely new repo).

## AI triage
The prior triage's claims all check out in the clone: the hub depends on `gerr-cosmos v1.1.0` directly (the old `danwt/gerr` is indirect-only), and the genesis-bridge data lives as a hub proto file that is duplicated in the RDK. The "new human reply" is danwt's 2024-06-14 comment — it was already folded into the current brief, and it does not settle the decision the issue is blocked on: danwt explicitly says he has no strong opinion, and the disagreement with omritoptix over where a shared package could live (given the circular-dependency objection) was never resolved by anyone. No human has answered the workforce's standing question of whether the issue is now obsolete.

## Research summary

Issue #931 (filed by `danwt`, June 2024) is not a bug report — it's a tech-debt tracking issue anticipating two debts the then-incoming genesis bridge would create: (1) the "messy gerrs" error package being used in both the hub and dymint, and (2) duplicated genesis-bridge data definitions ("json definitions") living in both the RDK and the hub. The body proposes moving the shared definitions into either the RDK or the Hub, accepting a cross-repo dependency rather than a new repo.

The comment thread is a design debate that never reached a decision. `omritoptix` objected that a hub↔rollapp dependency "could lead to circular dependencies which I'm definitely against" (the hub will eventually need rollapp DRSs for fraud proofs, while the rollapp would need the hub for the shared package). `danwt` replied that he has no strong opinion, that Michael was against adding more repos, but that he himself is "happy to add third/fourth/nth shared repos" and considers shared repos the status quo and "just the normal way of doing development" (citing that rdk, hub, dymint, and rollapp-evm all already depend on testify). That is an expression of ambivalence, not a decision, and the thread has had no human activity since 2024-06-14.

Verified read-only against the clone (`945d7361`, 2026-08-15) and the RDK repo:

- **Debt #1 is confirmed resolved.** The hub depends on `github.com/dymensionxyz/gerr-cosmos v1.1.0` — a dedicated shared error-package repo, i.e. exactly the "nth shared repo" approach danwt endorsed. The old `danwt/gerr` package survives only as an indirect dependency; no hub source file imports it.
- **Debt #2 evolved rather than persisted as described.** The genesis-bridge data is no longer ad-hoc JSON: the hub defines it via protobuf (`dymensionxyz/dymension/rollapp/genesis_bridge_data.proto`, generated into `x/rollapp/types`), and the RDK carries its own parallel proto. The two definitions are still duplicated across the repos and must stay wire-compatible, but this mirrors the status-quo posture danwt described, and the system has shipped and operated on it for two years.

## Root cause hypothesis

Not applicable in the defect sense — there is no bug. The issue remains open because the team never recorded a decision on the thread, while the debt it tracks was substantially resolved out-of-band: the shared-repo approach won in practice (`gerr-cosmos`, `sdk-utils`), and the remaining definition duplication became an intentional pair of wire-compatible proto files rather than an accident. The issue is stale.

## Areas involved

- Hub: `x/rollapp/types` genesis-bridge data definitions and the `gerr-cosmos` / `sdk-utils` dependencies in `go.mod`.
- RDK (separate repo, out of this clone's scope): `x/hub-genesis` and its parallel genesis-bridge proto definition.
- Cross-repo dependency topology between hub, RDK, and dymint — the actual subject of the unresolved debate.

## Acceptance criteria

There is no implementable fix to accept until the maintainers decide the issue's remaining scope. The only sensible outcomes are:

1. A maintainer closes the issue as resolved/stale, on the grounds that debt #1 is done (`gerr-cosmos`) and debt #2's duplication is the accepted status quo; or
2. A maintainer decides the duplicated genesis-bridge proto definitions should be consolidated, in which case a new, concretely-scoped issue is filed naming the chosen home (hub, RDK, or a new shared repo) — the exact point `omritoptix` and `danwt` disagreed on without resolution.

## Open questions

- For the maintainers (`danwt` / `omritoptix`): is this issue now obsolete given `gerr-cosmos` resolved the gerrs debt and the duplicated genesis-bridge proto definitions have been the accepted status quo for two years — or do you still want the definitions consolidated, and if so, into which repo given the circular-dependency objection on the thread?

**Comment by omritoptix** [comment]
I think this could lead to circular dependencies which I'm definitely against. 

e.g:
1. we will require rollapp DRSs in the hub (for fraud proof) - best to avoid until we'll have more elegant solution but as a first step maybe. 
2. the rollapp will require the hub (for this package) 

This could get messy fast imo.

**Comment by danwt** [comment]
Ok I dont' have a strong opinion
I do feel we need somewhere for 'shared' stuff
Michael was against adding more repos

I am happy to add third/fourth/nth.. shared repos
This is actually the status-quo: ie.g rdk, hub, dymint and rollapp-evm all depend on testify/
It's just the normal way of doing development

**Comment by omriai1** [comment]
Re-groomed with the update above, but it still needs a human:

- For the maintainers (`danwt` / `omritoptix`): is this issue now obsolete given `gerr-cosmos` resolved the gerrs debt and the duplicated genesis-bridge proto definitions have been the accepted status quo for two years — or do you still want the definitions consolidated, and if so, into which repo given the circular-dependency objection on the thread?

Answer in a new comment to try again. If posting from the manager's GitHub account, tag `/reuven`.

**Comment by omriai1** [comment]
_Re-groom bookkeeping — updated in place, safe to ignore._

**Comment by omriai1** [comment]
Re-groomed with the new answers, but it still needs a human:

- For the maintainers (`danwt` / `omritoptix`): is this issue now obsolete given `gerr-cosmos` resolved the gerrs debt and the duplicated genesis-bridge proto definitions have been the accepted status quo for two years — or do you still want the definitions consolidated, and if so, into which repo given the circular-dependency objection on the thread?

Answer in a new comment to try again. If posting from the manager's GitHub account, tag `/reuven`.

**Comment by omriai1** [comment]
## Research notes

- needs human judgment: decide whether the remaining proto duplication is accepted debt, or must be consolidated; if consolidated, choose Hub, RDK, or a shared repo.
- Debt one is resolved in practice. Commit `a2622758` and PR `#959` replaced local error helpers with `gerr-cosmos`. Current `go.mod:36` uses version `v1.1.0`; old `danwt/gerr` is indirect only at `go.mod:111`.
- The Hub still owns an active wire schema. `rollapp/genesis_bridge_data.proto` defines `GenesisBridgeData`; `query.proto` exposes `ValidateGenesisBridge`; and `genesis_bridge_data_validator.go` validates it.
- History shows continued ownership, not removal: commit `62fe71c7` and PR `#1564` moved validation into rollapp types and added the query. Later proto updates continued into `2025`.
- No repository evidence records approval of duplication or selects a canonical home. The issue thread instead records the unresolved cycle risk versus shared-repo tradeoff. The supplied closed issues do not settle it.
